docs/design/specs/Ecma-335-Augments.md

```diff
@@ -810,3 +820,84 @@ Bullet 2: It is valid to call a virtual method using `call` (rather than `callvi
 compile calls to “methods on `base`” (i.e., the statically known parent class) or to virtual static methods.
+
+## Covariant Return Types
+
+Covariant return methods is a runtime feature designed to support the [covariant return types](https://github.com/dotnet/csharplang/blob/master/proposals/csharp-9.0/covariant-returns.md) and [records](https://github.com/dotnet/csharplang/blob/master/proposals/csharp-9.0/records.md) C# language features implemented in C# 9.0.
+
+This feature allows an overriding method to have a return type that is different than the one on the method it overrides, but compatible with it. The type compability rules are defined in ECMA I.8.7.1.
+
+Example:
+```
+class Base
+{
+  public virtual object VirtualMethod() { return null; }
+}
+
+class Derived : Base
+{
+  // Note that the VirtualMethod here overrides the VirtualMethod defined on Base even though the 
+  // return types are not the same. However, the return types are compatible in a covariant fashion.
+  public override string VirtualMethod() { return null;}
+}
+```
+
+Covariant return methods can only be described through MethodImpl records, and as are only be applicable to methods on reference types. Methods on interfaces and value types are not supported. In addition the covariant override is only applicable for overriding methods also defined on a reference type.
+
+See [Implementation Design](../features/covariant-return-methods.md) for notes from the implementation in CoreCLR.
+
+Changes to the spec. These changes are relative to the 6th edition (June 2012) of the ECMA-335 specification published by ECMA available at:
+
+https://www.ecma-international.org/publications-and-standards/standards/ecma-335/
+
+### I.8.7.1 Assignment compatibility for signature types
+
+Add a new signature type relation.
+
+"A method signature type T is *covariant-return-compatible-with* with a method signature type U if and only if:
+1. The calling conventions of T and U shall match exactly, ignoring the distinction between static and instance methods (i.e., the this parameter, if any, is not treated specially).
+2. For each parameter type of P of T, and corresponding type Q of U, P is identical Q.
+3. For the return type P of T, and return type Q of U, Q is *compatible-with* P and is a *reference type* OR Q is identical to P.
+
+### II.10.3.2 The .override directive (page 147)
+
+Edit the first paragraph "The .override directive specifies that a virtual method shall be implemented (overridden), in this type, by a virtual method with a different name, but with a *covariant-return-compatible-with* signature (§I.8.7.1). This directive can be used to provide an implementation for a virtual method inherited from a base class, or a virtual method specified in an interface implemented by this type and all other. If not used to provide an implementation for a virtual method inherited from a base class the signature must be the identical."
```

Comment:
Nit - "the signature must be the identical" - as you know I'm not a native English speaker but should this rather be *the signature must be identical* or alternatively *the signature must be the same*?

Reply:
Sigh. My reading tends to gloss over that sort of issue, which is why a non-English speaker is actually quite helpful in this kind of review as you don't do that. Yes. the signature must be identical sounds right to me.